### PR TITLE
Reflect v0.15 schema changes

### LINKED ISSUE #27: Add support for protocol v6 schema changes
i.e. reflect https://github.com/hashicorp/terraform/pull/28055

### PR DESCRIPTION
**Pull request by radeksimko**
This reflects a change introduced a while ago in https://github.com/hashicorp/terraform/pull/20949 and also changes introduced as part of https://github.com/hashicorp/terraform/pull/28055, https://github.com/hashicorp/terraform/pull/27699 and https://github.com/hashicorp/terraform/pull/27793

The mentioned PRs do _not_ change anything about the underlying _syntax_ that couldn't be already expressed using the `SchemaAttribute.AttributeType`, however these changes allow the SDK and providers to communicate some additional metadata that cannot be communicated via "pure" `cty.Type`, such as sensitivity, or description for each individual attribute.

These new features are not in use by the SDK nor the built-in `terraform` provider yet (AFAIK), but this allows us to support it when the time comes.

Closes #27

### CODE REVIEW

**Comment by radeksimko** [comment]
Marking this as a draft temporarily until a version bump is decided/implemented per https://github.com/hashicorp/terraform/issues/28050#issuecomment-799405982

**Comment by radeksimko** [comment]
Version bump implemented in https://github.com/hashicorp/terraform/pull/28115

**Comment by radeksimko** [comment]
Given that this package mostly doesn't attempt to interpret the data, I think the job of deeper validation - such as whether `AttributeNestedType` and `AttributeType` conflict with each other should be left up to the consumer. I'm open to alternative ideas/opinions though.

File: schemas.go
Comment (vancluever) [inline schemas.go:41]: _Minor_ assuming that this is going to be logic that will need to be changed in the future.

**Comment by vancluever** [comment]
Some questions that I have to put in here for now and edit later as GH doesn't have draft comments yet 😛 (may update later)

q: Is `SchemaNestedAttributeType` live yet? Will integrations be expected to support both?
a: Does not seem to be yet.

q: Was MaxItems/MinItems supported elsewhere before?
a: Yes, in [`SchemaBlockType`](https://github.com/hashicorp/terraform-json/blob/0a260b947761224d457a0eb15d74af35a4c8293b/schemas.go#L154-L160).

q: What will integrations need to change?
a: You will need to start using `AttributeNestedType` for nested attributes and types.  [`NestedBlocks`](https://github.com/hashicorp/terraform-json/blob/0a260b947761224d457a0eb15d74af35a4c8293b/schemas.go#L103-L152) is unused in 0.2 schemas. That means you should check `AttributeNestedType` first to see if recursion is necessary for nested blocks before looking to `AttributeType` for non-nested attributes. If your integration supports both 0.1 and 0.2 you will need codepaths for both.

**Review by vancluever — APPROVED** [review]

File: schemas.go
Comment (radeksimko) [reply schemas.go:41]: Yes, I assume so, but from what I understand it most likely won't happen until the next major version of Terraform itself.

**Comment by radeksimko** [comment]
Your notes above look accurate.

FWIW naming-wise I think that calling the root thing a block was and is inaccurate, because "block" includes the header (with type and labels). Calling it a "body" would've been IMO better. This is also how HCL internally refers to these things. But even the docs call it a "block" apparently https://www.terraform.io/docs/cli/commands/providers/schema.html#block-representation

This might be something to address prior to v1 though, certainly not now and not in this PR.

**Comment by radeksimko** [comment]
btw. I don't seem to have write access and so I can't merge this PR - would you do that for me, please?

**Comment by vancluever** [comment]
> Your notes above look accurate.
> 
> FWIW naming-wise I think that calling the root thing a block was and is inaccurate, because "block" includes the header (with type and labels). Calling it a "body" would've been IMO better. This is also how HCL internally refers to these things. But even the docs call it a "block" apparently https://www.terraform.io/docs/cli/commands/providers/schema.html#block-representation
> 
> This might be something to address prior to v1 though, certainly not now and not in this PR.

Not opposed to this, I'll leave it to you 🙂 the rationale around most of the field _names_ was to ensure they matched what came in from the JSON, where possible.

**Pull request merged**
